dex/networks/zec/tx.go

```diff
@@ -0,0 +1,417 @@
+	ver, err := readUint32(r)
+	if err != nil {
+		return err
+	}
+
+	// overWintered := (ver & (1 << 31)) > 0
+	ver &= overwinterMask // Clear the overwinter bit
+	tx.Version = int32(ver)
+
+	if ver > 4 {
+		return fmt.Errorf("unsupported tx version %d > 4", ver)
+	}
+
+	if ver > 3 {
+		_, err := readUint32(r)
+		if err != nil {
+			return err
+		}
+	}
+
+	txInCount, err := wire.ReadVarInt(r, pver)
+	if err != nil {
+		return err
+	}
+
+	tx.TxIn = make([]*wire.TxIn, 0, txInCount)
+	for i := 0; i < int(txInCount); i++ {
+		ti := new(wire.TxIn)
+		if err := readTxIn(r, pver, int32(ver), ti); err != nil {
+			return err
+		}
+		tx.TxIn = append(tx.TxIn, ti)
+	}
+
+	txOutCount, err := wire.ReadVarInt(r, pver)
+	if err != nil {
+		return err
+	}
+
+	tx.TxOut = make([]*wire.TxOut, 0, txOutCount)
+	for i := 0; i < int(txOutCount); i++ {
+		to := new(wire.TxOut)
+		if err := readTxOut(r, pver, int32(ver), to); err != nil {
+			return err
+		}
+		tx.TxOut = append(tx.TxOut, to)
+	}
+
+	tx.LockTime, err = readUint32(r)
+	if err != nil {
+		return err
+	}
+
+	if ver > 3 {
+		tx.ExpiryHeight, err = readUint32(r)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Empty the rest of the buffer.
+	if ver < 2 {
+		return nil
+	}
+
+	var bindingSigRequired bool
+	if ver >= 4 {
+		// valueBalanceSpending
+		if _, err := readUint64(r); err != nil {
+			return fmt.Errorf("error reading valueBalanceSpending: %w", err)
+		}
+
+		if nSpendsSapling, err := wire.ReadVarInt(r, pver); err != nil {
+			return fmt.Errorf("error reading nSpendsSapling: %w", err)
+		} else if nSpendsSapling > 0 {
+			// vSpendsSapling - discard
+			bindingSigRequired = true
+			if _, err := io.ReadFull(r, make([]byte, nSpendsSapling*384)); err != nil {
+				return fmt.Errorf("error reading vSpendsSapling: %w", err)
+			}
+		}
+
+		if nOutputsSapling, err := wire.ReadVarInt(r, pver); err != nil {
+			return fmt.Errorf("error reading nOutputsSapling: %w", err)
+		} else if nOutputsSapling > 0 {
+			// vOutputsSapling - discard
+			bindingSigRequired = true
+			if _, err := io.ReadFull(r, make([]byte, nOutputsSapling*948)); err != nil {
+				return fmt.Errorf("error reading vOutputsSapling: %w", err)
+			}
+		}
+
+	}
+
+	if nJoinSplit, err := wire.ReadVarInt(r, pver); err != nil {
+		return fmt.Errorf("error reading nJoinSplit: %w", err)
+	} else if nJoinSplit > 0 {
+		// vJoinSplit - discard
+		sz := 1802 * nJoinSplit
+		if ver == 4 {
+			sz = 1698 * nJoinSplit
+		}
+		if _, err := io.ReadFull(r, make([]byte, sz)); err != nil {
+			return fmt.Errorf("error reading vJoinSplit: %w", err)
+		}
+		var joinSplitPubKey [32]byte
+		if _, err := io.ReadFull(r, joinSplitPubKey[:]); err != nil {
+			return fmt.Errorf("error reading joinSplitPubKey: %w", err)
+		}
+
+		var joinSplitSig [64]byte
+		if _, err := io.ReadFull(r, joinSplitSig[:]); err != nil {
+			return fmt.Errorf("error reading joinSplitSig: %w", err)
+		}
+	}
+
+	if ver == 4 && bindingSigRequired {
```

Comment:
Really not just `if bindingSigRequired {`?  Is it really intended that version 5 could set `bindingSigRequired = true`, not to hit this condition and read and discard these bytes?